File: backtracking/0051-n-queens.go
package backtracking

// https://leetcode.com/problems/n-queens/

func solveNQueens(n int) [][]string {
	solutions := make([][]string, 0)

	board := make([][]byte, n)
	for row := range board {
		board[row] = make([]byte, n)
		for col := range board[row] {
			board[row][col] = '.'
		}
	}

	cols := make([]bool, n)      // (row)
	negDiag := make([]bool, 2*n) // (row - col)
	posDiag := make([]bool, 2*n) // (row + col)

	var backtrack func(row int)

	backtrack = func(row int) {
		if row == n {
			solution := make([]string, n)
			for idx := range solution {
				solution[idx] = string(board[idx])
			}

			solutions = append(solutions, solution)

			return
		}

		for col := range n {
			if cols[col] || negDiag[n-row+col] || posDiag[row+col] {
				continue
			}

			board[row][col] = 'Q'
			cols[col] = true
			negDiag[n-row+col] = true
			posDiag[row+col] = true

			backtrack(row + 1)

			board[row][col] = '.'
			cols[col] = false
			negDiag[n-row+col] = false
			posDiag[row+col] = false
		}
	}
	backtrack(0)

	return solutions
}

File: backtracking/0051-n-queens.py
from typing import List
import unittest

# https://leetcode.com/problems/n-queens/

# python3 -m unittest backtracking/0051-n-queens.py

class Solution(unittest.TestCase):
    def solveNQueens(self, n: int) -> List[List[str]]:
        solutions = []
        board = [['.'] * n for _ in range(n)]
        cols = [False] * n # (col)
        negDiag = [False] * 2*n # (row-col)
        posDiag = [False] * 2*n # (row+col)
        def backtrack(row: int) -> None:
            if row == n:
                solutions.append(["".join(row) for row in board])
                return

            for col in range(n):
                if cols[col] or negDiag[n-row+col] or posDiag[row+col]:
                    continue

                board[row][col] = 'Q'
                cols[col] = True
                negDiag[n-row+col] = True
                posDiag[row+col] = True

                backtrack(row+1)

                board[row][col] = '.'
                cols[col] = False
                negDiag[n-row+col] = False
                posDiag[row+col] = False

        backtrack(0)
        return solutions

    def test(self) -> None:
        for n, expected in [
            (4, [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]),
            (1, [["Q"]]),
        ]:
            output = self.solveNQueens(n)
            self.assertListEqual(expected, output, f"expected: {expected}, output: {output}")
